src/margo-core.c

```diff
@@ -1585,20 +1586,32 @@ int __margo_internal_finalize_requested(margo_instance_id mid)
     return mid->finalize_requested;
 }
 
-void __margo_internal_incr_pending(margo_instance_id mid)
+int __margo_internal_incr_pending(margo_instance_id mid)
 {
-    if (!mid) return;
+    if (!mid) return -2;
+    if (mid->finalize_requested) return -1;
     ABT_mutex_lock(mid->pending_operations_mtx);
-    mid->pending_operations += 1;
+    while (mid->max_pending_operations
```

Comment:
I like the fix to the finalize_requested logic, can that be split out as a separate PR?

We can iterate on the question of how to best throttle RPC volume.  I still would like to unify it somehow at a lower level but it's not clear how yet.

Reply:
Ok I did another PR for this.

Regarding the problem of throttling, I asked Jerome and apparently if we make the RPC handler return HG_AGAIN it will be propagated to the sender. So we could use this as a mechanism to limit the number of on-going RPCs. That said, this mechanism would treat all the RPCs as equals, while we may want to limit only some RPCs (typically the long-running ones or the ones with large data transfers), or have different limits for different RPCs.

Reply:
maybe there could be a higher level scheduler that hooks into the rpc handlers and the argobots scheduler so that something has full visibility into that stuff and has the ability to influence both scheduling of handlers that have started and also shed load on those that haven't?